src/main/java/org/ibp/api/java/impl/middleware/program/validator/ProgramFavoriteValidator.java

```diff
@@ -0,0 +1,131 @@
+import org.generationcp.middleware.manager.ontology.daoElements.VariableFilter;
+import org.generationcp.middleware.pojos.Location;
+import org.generationcp.middleware.pojos.dms.ProgramFavorite;
+import org.ibp.api.exception.ApiRequestValidationException;
+import org.ibp.api.java.impl.middleware.common.validator.LocationValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.BindingResult;
+import org.springframework.validation.MapBindingResult;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Component
+public class ProgramFavoriteValidator {
+
+	@Autowired
+	private LocationValidator locationValidator;
+
+	@Autowired
+	private LocationDataManager locationDataManager;
+
+	@Autowired
+	private BreedingMethodService breedingMethodService;
+
+	@Autowired
+	private OntologyVariableDataManager ontologyVariableDataManager;
+
+	@Autowired
+	private ProgramFavoriteService programFavoriteService;
+
+	public void validateAddFavorites(final String programUUID, final ProgramFavoriteRequestDto programFavoriteRequestDtos) {
+		final BindingResult errors = new MapBindingResult(new HashMap<>(), ProgramFavoriteRequestDto.class.getName());
+		final Set<Integer> entityIds = new HashSet<>(programFavoriteRequestDtos.getEntityId());
+		final ProgramFavorite.FavoriteType favoriteType = programFavoriteRequestDtos.getFavoriteType();
+
+		if (CollectionUtils.isEmpty(programFavoriteRequestDtos.getEntityId())) {
+			errors.reject("program.favorite.entity.list.id.required", "");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		final List<ProgramFavorite> favorites = this.programFavoriteService.getProgramFavorites(programUUID, favoriteType, entityIds);
+
+		if (!favorites.isEmpty()) {
+			List<Integer> favoritesIds = favorites.stream().map(ProgramFavorite::getEntityId).collect(Collectors.toList());
+			errors.reject("program.favorite.already.exists", new String[] {
+				String.join(",", Arrays.toString(favoritesIds.toArray())), favoriteType.getName()}, "");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+
+		}
+
+		switch (favoriteType) {
+			case LOCATION:
+				this.validateLocationId(errors, programUUID, entityIds);
+				break;
+			case METHOD:
+				this.validateMethodId(errors, programUUID, entityIds);
+				break;
+			case VARIABLE:
+				this.validateVariableId(errors, programUUID, entityIds);
+				break;
+			default:
+				errors.reject("program.favorite.type.not.identified", new String[] {favoriteType.getName()}, "");
+				throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+	}
+
+	private void validateVariableId(final BindingResult errors, final String programUUID, final Set<Integer> variableIds) {
```

Comment:
I think the next 3 validators should be located in their respective classes because they can be reused. Also, the property message can be unique removing the program.favorite section 
In all cases the function name must be renamed to Ids as the function is validating a set